### PR TITLE
Add TypeScript signatures(#22)

### PR DESCRIPTION
**Pull request by EvgeniyKumachev**
**What**:
Added TypeScript signatures

**Why**:

**How**:

**Checklist**:

* [x] Documentation
* [x] Tests
* [x] Ready to be merged 
* [x] Added myself to contributors table

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * This function decides what message to show based on the icon (returned from getIcon prop) and
+     * the current state of the component.
+     */
+    getMessage?: (icon: string, state: LoadingState) => string
+    /**
+     * This function is called as soon as the component enters the viewport and is used to generate urls
+     * based on width and format if props.srcSet doesn't provide src field.
+     */
+    getUrl?: () => string
```
Comment (stereobooster) [inline index.d.ts]: `getUrl?: (srcType) => string`

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  export interface ImageProps {
+    /**
+     * This function decides what icon to show based on the current state of the component.
+     */
+    getIcon?: (state: LoadingState) => string
+    /**
+     * This function decides what message to show based on the icon (returned from getIcon prop) and
+     * the current state of the component.
+     */
+    getMessage?: (icon: string, state: LoadingState) => string
```
Comment (stereobooster) [inline index.d.ts]: icons - can be expressed as union type

**Comment by stereobooster** [comment]
Type definitions are good. I wonder if we can add some kind of test, to make sure this definitions stays valid? 🤔 Maybe sample typescript project, like [they do in downshift](https://github.com/paypal/downshift/blob/995501349e1cfdd933fae29797cd0dae4d1bba34/package.json#L19)

**Comment by EvgeniyKumachev** [comment]
Haven't tried writing tests for typings yet. I'll see what i can do

**Comment by EvgeniyKumachev** [comment]
I've taken [downshift](https://github.com/paypal/downshift) for example and added `test:ts` script in validation. Maybe typings tests should have separate directory but for now i've put in in `src/__tests__`

File: package.json [hunk 3]
```diff
@@ -13,7 +14,8 @@
     "lint": "kcd-scripts lint",
     "test": "kcd-scripts test",
     "test:update": "npm test -- --updateSnapshot --coverage",
-    "validate": "kcd-scripts validate",
+    "test:ts": "tsc --noEmit -p ./tsconfig.json",
+    "validate": "kcd-scripts validate test:ts",
```
Comment (stereobooster) [inline package.json]: test:ts && kcd-scripts validate

File: .all-contributorsrc [hunk 4]
```diff
@@ -63,6 +63,15 @@
...
         "doc"
       ]
+    },
+    {
+      "login": "EvgeniyKumachev",
+      "name": "Evgeniy Kumachev",
+      "avatar_url": "https://avatars3.githubusercontent.com/u/5207796?v=4",
+      "profile": "https://github.com/EvgeniyKumachev",
+      "contributions": [
+        "doc"
```
Comment (stereobooster) [inline .all-contributorsrc]: I would say this is code, not a doc. But I'm ok with doc too

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+  "compilerOptions": {
+    "jsx": "react",
+    "noUnusedLocals": true,
+    "strict": false,
```
Comment (stereobooster) [inline tsconfig.json]: shouldn't it be true?

Comment (EvgeniyKumachev) [reply tsconfig.json]: I'm importing icons and they are implicitly `any` because i didn't write types for them

Comment (stereobooster) [reply tsconfig.json]: `{ 'noicon':  () => null }` should work (I guess)

File: src/__tests__/basic.test.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+export default class Application extends React.Component<Props, State> {
+  render() {
+    return (
+      <IdealImage
+        theme={{placeholder: {filter: 'blur(3px)'}}}
+        srcSet={[{src: 'some-src.jpg', width: 3500}]}
+        placeholder={{color: '#FFFFFF'}}
+        shouldAutoDownload={() => true}
+        loader="image"
+        getUrl={srcType => `some-src-${srcType}.jpg`}
```
Comment (stereobooster) [inline src/__tests__/basic.test.tsx]: `some-src-${srcType.width}.jpg`?

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  format?: 'webp' | 'jpeg'
+}
+
+type ThemeKey = 'placeholder' | 'img' | 'icon' | 'noscript'
+
+export interface ImageProps {
+  /**
+   * This function decides what icon to show based on the current state of the component.
+   */
+  getIcon?: (state: LoadingState) => string
```
Comment (stereobooster) [inline index.d.ts]: `getIcon?: (state: LoadingState) => IconKey `

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * https://github.com/stereobooster/react-ideal-image/blob/master/introduction.md#lqip
+   */
+  placeholder: {color: string} | {lqip: string}
+  /**
+   * This function decides if image should be downloaded automatically. The default function
+   * returns false for a 2g network, for a 3g network it decides based on props.threshold
+   * and for a 4g network it returns true by default.
+   */
+  shouldAutoDownload?: () => boolean
```
Comment (stereobooster) [inline index.d.ts]:
```ts
shouldAutoDownload?: ({
  connection?: 'slow-2g' | '2g' | '3g' |  '4g'
  size?: number,
  threshold?: number,
  possiblySlowNetwork?: boolean,
}) => boolean
```

File: src/__tests__/basic.test.tsx [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+export default class Application extends React.Component<Props, State> {
+  render() {
+    return (
+      <IdealImage
+        theme={{placeholder: {filter: 'blur(3px)'}}}
+        srcSet={[{src: 'some-src.jpg', width: 3500}]}
+        placeholder={{color: '#FFFFFF'}}
+        shouldAutoDownload={() => true}
+        loader="image"
+        getUrl={srcType => srcType.src}
```
Comment (stereobooster) [inline src/__tests__/basic.test.tsx]: this should throw an error, because `srcType.src` is `string?`, where is  `getUrl` is `(...) => string`

Comment (stereobooster) [reply src/__tests__/basic.test.tsx]: Actually it does

```
src/__tests__/basic.test.tsx:16:9 - error TS2326: Types of property 'getUrl' are incompatible.
  Type '(srcType: SrcType) => string | undefined' is not assignable to type '((srcType: SrcType) => string) | undefined'.
    Type '(srcType: SrcType) => string | undefined' is not assignable to type '(srcType: SrcType) => string'.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.
16         getUrl={srcType => srcType.src}
```

**Pull request merged**